llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -966,6 +966,41 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
                 I, 1, (DemandedMask & ~LHSKnown.Zero).zextOrTrunc(MaskWidth)))
           return I;
 
+        // Combine:
+        // (ptrmask (getelementptr i8, ptr p, imm i), imm mask)
+        //   -> (ptrmask (getelementptr i8, ptr p, imm (i & mask)), imm mask)
+        // where only the low bits known to be zero in the pointer are changed
+        Value *InnerPtr;
+        uint64_t GEPIndex;
+        uint64_t PtrMaskImmediate;
+        if (match(I, m_Intrinsic<Intrinsic::ptrmask>(
+                         m_GEP(m_Value(InnerPtr), m_ConstantInt(GEPIndex)),
+                         m_ConstantInt(PtrMaskImmediate)))) {
+          auto *GEP = cast<GetElementPtrInst>(II->getArgOperand(0));
+
+          if (GEP->getSourceElementType() == Type::getInt8Ty(I->getContext())) {
+            unsigned Log2Align = llvm::Log2(InnerPtr->getPointerAlignment(DL));
+            uint64_t PointerAlignBits = (uint64_t(1) << Log2Align) - 1;
+
+            uint64_t HighBitsGEPIndex = GEPIndex & ~PointerAlignBits;
+            uint64_t MaskedLowBitsGEPIndex =
+                GEPIndex & PointerAlignBits & PtrMaskImmediate;
+
+            uint64_t MaskedGEPIndex = HighBitsGEPIndex | MaskedLowBitsGEPIndex;
+
+            if (MaskedGEPIndex != GEPIndex) {
+              Type *I64 = Type::getInt64Ty(I->getContext());
+              auto MaskedGEP =
```

Comment:
```suggestion
              Value *MaskedGEP =
```
I believe you also need to set the Builder insertion point here, otherwise the GEP may be created in the wrong place.

Reply:
Nice catch on builder, yes